File: 85-maximal-rectangle/maximal-rectangle.cpp
// Time Complexity: O(m * n), where m = number of rows, n = number of columns
// Space Complexity: O(n)

// Approach:
// • Treat each row as the base of a histogram: compute heights of consecutive '1's up to that row.
// • For each row’s histogram, compute the largest rectangle area (using stack in O(n) time).
// • Keep track of the maximum rectangle found across all rows.

class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if (matrix.empty() || matrix[0].empty()) return 0;
        int m = matrix.size(), n = matrix[0].size();
        vector<int> heights(n, 0);
        int maxArea = 0;

        for (int i = 0; i < m; ++i) {
            // update histogram heights
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] == '1') {
                    heights[j] += 1;
                } else {
                    heights[j] = 0;
                }
            }
            // compute largest rectangle area in this histogram
            maxArea = max(maxArea, largestRectangleArea(heights));
        }
        return maxArea;
    }

private:
    int largestRectangleArea(const vector<int>& heights) {
        int n = heights.size();
        stack<int> st;
        int maxArea = 0;
        for (int i = 0; i <= n; ++i) {
            int currentHeight = (i == n ? 0 : heights[i]);
            while (!st.empty() && currentHeight < heights[st.top()]) {
                int h = heights[st.top()];
                st.pop();
                int width = st.empty() ? i : i - st.top() - 1;
                maxArea = max(maxArea, h * width);
            }
            st.push(i);
        }
        return maxArea;
    }
};

File: 85-maximal-rectangle/maximal-rectangle.java
/**
 * Approach:
 * 1. We transform the matrix into a series of histogram heights.
 *    For each row:
 *      - If cell == '1', increase height at that column by 1.
 *      - If cell == '0', reset height at that column to 0.
 * 2. For each row’s heights[] array, compute the largest rectangle in the histogram.
 *    Use a monotonic increasing stack of column indices:
 *      - Push indices while current height is >= stack top’s height.
 *      - When we see a smaller height, pop from the stack and compute area:
 *          height = heights[popped],
 *          width = current index − stack top after pop − 1
 *      - Repeat until stack is valid.
 * 3. Track the maximum area seen across all rows.
 * 4. Time is O(rows * cols), as each histogram computation is linear with a stack.
 */

// Time Complexity: O(m * n)  Space Complexity: O(n)
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0) return 0;
        int m = matrix.length, n = matrix[0].length;
        int[] heights = new int[n];
        int maxArea = 0;

        for (int i = 0; i < m; i++) {
            // Build histogram heights for this row
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    heights[j] += 1;
                } else {
                    heights[j] = 0;
                }
            }
            maxArea = Math.max(maxArea, largestRectangleArea(heights));
        }
        return maxArea;
    }

    private int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0, n = heights.length;
        for (int i = 0; i <= n; i++) {
            int curHeight = (i == n ? 0 : heights[i]);
            while (!stack.isEmpty() && curHeight < heights[stack.peek()]) {
                int height = heights[stack.pop()];
                int widthRightBoundary = i-1;
                int widthLeftBoundary = stack.empty() ? 0 : stack.peek()+1;
                int width = widthRightBoundary-widthLeftBoundary+1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }
        return maxArea;
    }
}
